File: 538.把二叉搜索树转换为累加树.py
#
# @lc app=leetcode.cn id=538 lang=python3
#
# [538] 把二叉搜索树转换为累加树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def dfs(node):
            if not node:
                return
            dfs(node.right)
            nonlocal sum
            sum+=node.val
            node.val=sum
            dfs(node.left)
        sum=0
        dfs(root)
        return root
# @lc code=end

# Solution1:
# class Solution:
#     def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
#         def dfs(node):
#             if not node:
#                 return
#             dfs(node.right)
#             nonlocal sum
#             sum+=node.val
#             node.val=sum
#             dfs(node.left)
#         sum=0
#         dfs(root)
#         return root

File: 538.把二叉搜索树转换为累加树.java
/*
 * @lc app=leetcode.cn id=538 lang=java
 *
 * [538] 把二叉搜索树转换为累加树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    int sum = 0;

    public TreeNode convertBST(TreeNode root) {
        if (root == null) {
            return null;
        }
        convertBST(root.right);
        sum += root.val;
        root.val = sum;
        convertBST(root.left);
        return root;
    }

}
// @lc code=end

// Solution2:
// class Solution {
// int sum = 0;

// public TreeNode convertBST(TreeNode root) {
// Deque<TreeNode> deque = new LinkedList<>();
// TreeNode node = root;
// while (!deque.isEmpty() || node != null) {
// while (node != null) {
// deque.push(node);
// node = node.right;
// }
// node = deque.pop();
// sum += node.val;
// node.val = sum;
// node = node.left;
// }
// return root;
// }
// }